### PR TITLE
[React] Fix optionality of inferred props

### LINKED ISSUE #33742: [@types/react] Destructured props inferred as required
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @Asana @AssureSign @johnnyreilly @bbenezech @pzavolinsky @digiguru @ericanderson @tkrotoff @DovydasNavickas @onigoetz @theruther4d @guilhermehubner @ferdaber @jrakotoharisoa @pascaloliv @hotell @franklixuefei @Jessidhia @pshrmn @threepointone @saranshkataria 

I'm working on migrating a React project from JS to TS. I keep running into this issue when rendering JS components from TS. If the JS component is a functional component with the props destructured, TS considers all props required:
```javascript
// B.js
import React from 'react'
import PropTypes from 'prop-types'

const B = ({ content }) => <div>{content}</div>

B.propTypes = {
  content: PropTypes.string,
}

export default B
```

```typescript
// A.tsx
import React from 'react'
import B from './B'

// Property 'content' is missing in type '{}' but required in type '{ content: string; }'.ts(2741)
const A = () => <B />

export default A
```

If `B` is defined as `const B = props => <div>{props.content}</div>`, `content` is correctly inferred as optional. Ideally, I would like to be able to consume such JS components without modification, though.

I tried to dig into the issue, but I don't fully understand where this inference is occurring; whether it's in the definition of `ReactManagedAttributes`/`LibraryManagedAttributes`, TypeScript's handling of `LibraryManagedAttributes`, or `InferProps` in `@types/prop-types`.

### PR DESCRIPTION
**Pull request by matt-tingen**
Please fill in this template.

-   [x] Use a meaningful title for the pull request. Include the name of the package modified.
-   [X] Test the change in your own code. (Compile and run.)
-   [X] Add or edit tests to reflect the change. (Run with `npm test`.)
-   [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
-   [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
-   [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:

-   [x] Provide a URL to documentation or source code which provides context for the suggested changes: Fixes #33742
-   [ ] ~~Increase the version number in the header if appropriate.~~
-   [ ] ~~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.~~

This change ensures that the optional modifier (`?`) is copied when using inferred properties from `propTypes`.

### CODE REVIEW

File: types/react/index.d.ts [hunk 1]
```diff
@@ -2732,16 +2732,18 @@ declare namespace React {
 // so boolean is only resolved for T = any
 type IsExactlyAny<T> = boolean extends (T extends never ? true : false) ? true : false;
 
+// Pick properties for which the value is `any`
+type PickAny<T> = Pick<T, { [K in keyof T]: IsExactlyAny<T[K]> extends true ? K : never }[keyof T]>;
```
Comment (ferdaber) [inline types/react/index.d.ts]: It looks like you're only using `PickAny` for the keys, right? Is the outer `Pick` needed then?
```ts
type ExactlyAnyPropertyKeys<T> = { [K in keyof T]: IsExactlyAny<T[K]> extends true ? K : never }[keyof T]
```

File: types/react/index.d.ts [hunk 2]
```diff
@@ -2732,16 +2732,18 @@ declare namespace React {
...
-type MergePropTypes<P, T> = IsExactlyAny<P> extends true ? T : ({
-    [K in keyof P]: IsExactlyAny<P[K]> extends true
-        ? K extends keyof T
-        ? T[K]
-        : P[K]
-        : P[K]
-} & Pick<T, Exclude<keyof T, keyof P>>);
+type MergePropTypes<P, T> = IsExactlyAny<P> extends true ? T : string extends keyof P ? P : (
+    // From declared props, pick properties which are either not `any` or are missing on `propTypes`.
+    & Pick<P, Exclude<keyof P, keyof PickAny<P>> | Exclude<keyof P, keyof T>>
```
Comment (ferdaber) [inline types/react/index.d.ts]: Maybe introducing another intermediate type would help make the code more clear?
```ts
type NotExactlyAnyPropertyKeys<T> = Exclude<keyof T, ExactlyAnyPropertyKeys<T>>
type MergePropTypes<P, T> = IsExactlyAny<P> extends true ? T : string extends keyof P
  ? P
    // properties that are not exactly any from props
  : & Pick<P, NotExactlyAnyPropertyKeys<P>>
    // properties that are left over from inferred propTypes
    & Pick<T, Exclude<keyof T, NotExactlyAnyPropertyKeys<P>>>
    // any other properties left behind
    & Pick<P, Exclude<keyof P, keyof T>>
```

**Comment by ferdaber** [comment]
Thanks for doing this, by the way! Great work!

**Comment by matt-tingen** [comment]
@ferdaber Thanks for the comments! I'll take a deeper look and update the PR tonight or tomorrow.

In regards to the tests, I was getting similar memory issues when running locally. Is this a common issue with the tests or could it indicate a problem with my changes?

**Comment by ferdaber** [comment]
It's an unfortunate issue, you'll need to run the tests locally (just close everything out and save all your ram :P), and if it all works you can post here and one of the maintainers will manually merge for you.

**Comment by matt-tingen** [comment]
@ferdaber I pushed a change to clean up `MergePropTypes` based on your suggestions.

Unfortunately, I have not been able to complete a test run on my laptop even from a cold boot. Would you mind doing a test run for me?

**Pull request closed without merging**